crypto/addresses/src/lib.rs

```diff
@@ -0,0 +1,144 @@
+use std::fmt::{Display, Formatter};
+
+mod bech32;
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub enum AddressError {
+    InvalidPrefix(String),
+    MissingPrefix,
+    DecodingError(char),
+    BadChecksum,
+}
+
+impl Display for AddressError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Address decoding failed: {}",
+            match self {
+                Self::InvalidPrefix(prefix) => format!("Invalid prefix {}", prefix),
+                Self::MissingPrefix => "Prefix is missing".to_string(),
+                Self::BadChecksum => "Checksum is invalid".to_string(),
+                Self::DecodingError(c) => format!("Invalid character {}", c),
+            }
+        )
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum Prefix {
+    Mainnet,
+    Testnet,
+    Devnet,
+}
+
+impl Prefix {
+    fn as_str(&self) -> &'static str {
+        match self {
+            Prefix::Mainnet => "kaspa",
+            Prefix::Testnet => "kaspatest",
+            Prefix::Devnet => "kaspadev",
+        }
+    }
+}
+
+impl Display for Prefix {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+impl TryFrom<&str> for Prefix {
+    type Error = AddressError;
+
+    fn try_from(prefix: &str) -> Result<Self, Self::Error> {
+        match prefix {
+            "kaspa" => Ok(Prefix::Mainnet),
+            "kaspatest" => Ok(Prefix::Testnet),
+            "kaspadev" => Ok(Prefix::Devnet),
+            _ => Err(AddressError::InvalidPrefix(prefix.to_string())),
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct Address {
+    pub prefix: Prefix,
+    pub payload: Vec<u8>,
+    pub version: u8,
+}
+
+impl From<Address> for String {
+    fn from(address: Address) -> Self {
+        format!("{}:{}", address.prefix, address.encode_payload())
+    }
+}
+
+impl TryFrom<String> for Address {
+    type Error = AddressError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        match value.split_once(':') {
+            Some((prefix, payload)) => Self::decode_payload(prefix.try_into()?, payload),
+            None => Err(AddressError::MissingPrefix),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Can you please add all the tests from here: https://github.com/kaspanet/kaspad/blob/54fbfadf84ab43cd4eff7df131ebd18b50bdd882/util/bech32/bech32_test.go

Reply:
I've added the test, but to keep then intact, I had to leave test-only prefixes in the Prefix enum:
```
#[derive(PartialEq, Eq, Clone, Debug)]
pub enum Prefix {
    Mainnet,
    Testnet,
    Devnet,
    #[cfg(test)]
    A,
    #[cfg(test)]
    B,
}
```
The alternative would be to recalculate the checksum, but the point is to check my calculations

Tell me if you want me to change the prefixes